tests/lang/s02_core/s02_procedures/t03_overload_core_bugs/t03_user_typeclass_is_not_a_proper_alias.nim

```diff
@@ -0,0 +1,24 @@
+discard """
+description: '''
+user-defined typeclass breaks overload resolution.
+
+Moving `A or B` in a separate generic typeclass fails overload
+resulution. `type Q = A or B` is not the same as `A or B` written
+directly in the proc parameter
+'''
+
+knownIssue: "https://github.com/nim-lang/Nim/issues/1385"
+knownIssue: "https://github.com/nim-lang/Nim/issues/8435 - related"
+
+errmsg: "Error: type mismatch: got <float, int>"
+"""
+
+type Typeclass = int or float # Adding distinct doesn't help
+
+proc y(a: int or float; b: int or float) = discard
+y(float(2.0), int(2))
+
+proc x(a: Typeclass; b: Typeclass) = discard
```

Comment:
I think you have to add `distinct` to the type name in the proc definition, meaning the type parameters are distinct -- it's a different meaning.

like so:
```nim
proc x(a: Typeclass; b: distinct Typeclass) = discard
```

See the implicit generics section of the manual: https://nim-lang.org/docs/manual.html#generics-implicit-generics

I'm not sure how I feel about this though, I'm very much on the fence with a slight lean towards the current behaviour being correct with implicit generics.

Reply:
I think that

```nim
proc impl(a: A or B, b: A or B) ...
```

and

```nim
type Alias = A or B
proc impl(a: Alias, b: Alias) ...
```

should function identically. If someone wanted to constrain both arguments to the specific "both A" or "both B" they could've written 

```nim
type Alias = int or float
proc impl[T: Alias](a: T, b: T) = discard

impl(12.0, 2.0)
```

Reply:
Intuitively, substituting `A = <expr>` in place of the `<expr>` should work the same way. `A = <modifier> <expr>` is different (like `A = distinct <expr>` etc.), but I don't think `A = distinct int or float` makes a lot of sense.

Reply:
> See the implicit generics section of the manual: nim-lang.org/docs/manual.html#generics-implicit-generics

I guess it makes sense, technically ... well, in that case I should probably consider that example works "according to the spec".

Found some interesting examples of the `distinct` typeclass feature

```nim
template `==`*(a, b: distinct (string|StringOfJson)): bool =
```

Reply:
I think we'd have to play with the implemtnation a bit because here the identifier `Alias` is doing two things:
1. referring to a specific type, in this case a typeclass, which makes the proc implicitly generic
2. acting as the name `T` is in the second example you wrote

There are some subtleties that'll likely show up by working through the implementation, one example I can think of that's tricky is that in a generic proc you can refer to a type parameter like `Alias` or `T` in the body, so here is a case of an inner proc:

```nim
proc impl(a: Alias, b: Alias): Alias =
  proc identity(i: Alias): Alias = i
  identity(b)

let c = impl(1, "string")
```

Nim semantically analyses, due to metaprogramming/sanity reasons, in a forward direction, from leaf to root. So the compiler will:
1. have a skeleton definition of the generic proc against the symbol `impl`
2. then find the call to `impl`, it'll fix the two types for the generic instance
3. in the generic instance, we either have an inner generic proc or do we have a fixed/known type?

So it's ambiguous, we can make a choice, but not sure we have enough information to choose which way is better.

Issues like the following come to mind with point 3:
a. `Alias` is only known if we treat it as a type parameter name, but 
b. if it's also generic, then ok, I don't think that results in any codegen issues; but how do I say I don't mean generic
c. if we use `impl[T: Alias]` then if a is passed in a more precise sub-type of either `int` or `string` it'll fixate on that, and force `b`'s type to narrow more than the `Alias`

although, now that I think about it, 'c' might be a bug already. 😬

Reply:
```nim
proc p(a: Table, b: distinct Table)

# is roughly the same as:

proc p[Key, Value, KeyB, ValueB](a: Table[Key, Value], b: Table[KeyB, ValueB])
```

Do you know what "roughly the same" means *exactly* in this case?

Reply:
> I'm not sure how I feel about this though, I'm very much on the fence with a slight lean towards the current behaviour being correct with implicit generics.

I now think current behavior makes sense, so we need to leave it as is, and I will turn the example from "bug" into "test the error"

Reply:
> ```nim
> proc p(a: Table, b: distinct Table)
> 
> # is roughly the same as:
> 
> proc p[Key, Value, KeyB, ValueB](a: Table[Key, Value], b: Table[KeyB, ValueB])
> ```
> 
> Do you know what "roughly the same" means _exactly_ in this case?

it means that the compiler sorta kinda does that transform, but significant details and all that differ.

Reply:
> > I'm not sure how I feel about this though, I'm very much on the fence with a slight lean towards the current behaviour being correct with implicit generics.
> 
> I now think current behavior makes sense, so we need to leave it as is, and I will turn the example from "bug" into "test the error". I need to read more into distinct generics, they seem to be very weird at places, and not really widely used.

I've become _less_ certain, here are the things I have more questions about, maybe @beef331 has thoughts.

Here are the various pieces I'm thinking of:
* `Alias` the type parameter name
* `Alias`, the name of a type or type alias in the type section
  * this is sorta ambiguous and very subtle, but is it the name of a type or an alias? current implementation leans alias
* type or type value, in the type section `Alias` has `typeclass2[int, string]`
  * we could treat it as `namedtypeclass2["Alias", int, string]` (see alias vs type ambiguity above) but currently the implementation treats it like `typeclass2[int, string]`... mostly
* type parameter value, the type value we bind to a given type parameter name, applying an unbound type parameter results in generics, while applying a bound type parameter resolves it and reduces the genericity

Some other misc thoughts, mostly settled:
* pretty sure it isn't, but is `distinct int or string` ambiguous? is it `distinct (int or string)` or `(distinct int) or string`; precedence rules wise, distinct is kinda like a "type call" so that happens after the type expression `int or string`
* distinct on a type parameter I believe has to apply to the type parameter name, so a type symbol operation, because if it applied to the type value it'd break the subtype relation

Things are still a bit off:
```nim
proc foo(a, b: int) = # `a` and `b` are both int
  discard

proc foo(a, b: int or string) = # `a` and `b` are which one of `int` or `string` don't have to be the same
  discard

type
  Base = object of RootObj
  Sub = object of Base

proc foo(a, b: Base) = # if I pass in a `Sub` for `a` it does _not_ fix `b`'s type
  discard

type Alias = int or string

proc foo(a, b: Alias) = # this issue

type Other = object of Base

proc foo(a, b: Base) = # also fine to pass in any combination of `Base, Sub, or Other`

proc foo(a, b: distinct Alias) = # works

proc foo(a, b: distinct int or string) = # doesn't work
```

My conclusion right now is that we stay close-ish to the spec, but the above is confusion, type alias vs type name, type parameter name and substitution, and type operators need to work more consistently.

Reply:
It's very counter intuitive, but it's also nice when you need it.  The inheritance version probably should be bound identically, instead of just allowing any subtype, since an inheritable type is practically a constraint of anything that comes from it.

Should be noted you do not need `a: T, b: distinct T` the following works:
```nim
proc x(a, b: distinct SomeNumber) = echo typeof(a), " ", typeof(b)
x(10, 20)
x(10f, 20)
x(10u8, 20d)
```

Reply:
It is possible to pass multiple different subtypes to the varargs, and given mechanics of how the pass is performed (only embedded supertype part is actually passed) it makes sense that `a, b: Base` allows for `Derived1(), Derived2()`